### PR TITLE
Setting up Data Validation classes

### PR DESCRIPTION
**Pull request by Shom770**
Creating year-agnostic and year-specific Data Validation classes.

### CODE REVIEW

File: backend/data_validation/base_data_val.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
+import collections
+import json
+import statistics
+
+import falcon_alliance
+import numpy as np
+import yaml
+from utils import *
```
Comment (amanjaiman) [inline backend/data_validation/base_data_val.py]: Generally it's good practice to import only the functionality you need from a library (unless it's A LOT) of things. It also makes it easier to identify exactly which library a certain function/class is coming from.

See this for example: https://stackoverflow.com/a/2386740

File: backend/data_validation/data_val_2022.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
+import collections
+import json
+from operator import itemgetter
+
+from base_data_val import BaseDataValidation
+from utils import *
```
Comment (amanjaiman) [inline backend/data_validation/data_val_2022.py]: See previous comment about this

File: backend/data_validation/base_data_val.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+            with open(
+                self.config.get("path_to_match_schedule", "../data/match_schedule.json")
+            ) as file:
+                self.match_schedule = json.load(file)
+        except FileNotFoundError:  # We want to ignore if it doesn't exist because get_match_schedule() will create it.
+            pass
+
+        # Setting up FalconAlliance (our connection to TBA) and retrieving match schedule if it doesn't exist
+        self.api_client = falcon_alliance.ApiClient(
+            api_key="6lcmneN5bBDYpC47FolBxp2RZa4AbQCVpmKMSKw9x9btKt7da5yMzVamJYk0XDBm"  # for testing purposes
```
Comment (amanjaiman) [inline backend/data_validation/base_data_val.py:46]: Not sure how strict we want to be with this. Usually you don't want to expose these to public so it's recommended to put it in your environment variables and read from there.

File: backend/data_validation/utils.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    Checks if the match key is valid.
+
+    :param key: Match key as a string (eg 'qm1')
+    :return: Boolean representing whether or not a match key is valid.
+    """
+    # Check if match key is valid via regex
+    match_key_format = re.compile(
+        r"(qm[1-9][0-9]{0,3})|(qf[1-4]m[1-3])|(sf[1-2]m[1-3])|(f[1]m[1-3])"
+    )
+    return bool(re.fullmatch(match_key_format, key))
```
Comment (amanjaiman) [inline backend/data_validation/utils.py]: Not sure if this throws an error if key is null - idk if fullmatch handles it or if you should, though good practice to check if input params are null so you can throw custom errors and user knows what went wrong

File: backend/data_validation/__main__.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import yaml
+from data_val_2022 import DataValidation2022
+
+validation_by_year = {2022: DataValidation2022}
+path_to_config = "config.yaml"
```
Comment (amanjaiman) [inline backend/data_validation/__main__.py:5]: Just a note: if this config file  is used in other files as a var, then it'll be better to create a constants file and add it in there. Then we use PATH_TO_CONFIG everywhere it's used

**Review by amanjaiman — COMMENTED** [review]
Overall good, few "best practice" changes I think can be made.

File: backend/data_validation/base_data_val.py (hunk 3 above)
Comment (sswadkar) [reply backend/data_validation/base_data_val.py:46]: github env variables are funky,, worth figuring out for the future but for strictly testing purposes it's fine imo. tba api keys are the only api keys i would recommend exposing at all (even then shouldn't expose once we release the app). a lot of teams expose tba api keys bc it's a lot of effort to hide as well.

File: requirements.txt [hunk 6]
```diff
@@ -1,15 +1,50 @@
...
+async-timeout==4.0.2
+attrs==22.1.0
 autopep8==1.7.0
 cfgv==3.3.1
+charset-normalizer==2.1.1
+contourpy==1.0.6
+coverage==6.4.4
+cycler==0.11.0
 distlib==0.3.6
+falcon-alliance==0.4.4
```
Comment (AlphaPranav9102) [inline requirements.txt:13]: @Shom770 1.0.0?

**Review by AlphaPranav9102 — COMMENTED** [review]
W manz

**Pull request merged**